Configurable stratum diff, block refresh, and extranonce

rdugan opened the pull request:
Three new configurable features, with defaults set to previous values:
 - stratum diff to reduce network traffic and server load.  Instantiated at the client level (rather than global) to allow easier transition to vardiff in the future if desired.
 - block refresh time, also to reduce network traffic and server load (as well as log spam)
 - extranonce support, to partition nonce space across multiple clients, preventing work overlap in case mining client does not randomize nonce

Also cleaned up CLI output a bit, including block counting.

c0sco commented:
This has been working great for me, and closes #43.

onemorebsmith commented on src/kaspastratum/hasher.go:
```diff
@@ -13,15 +13,46 @@ import (
+// static value definitions to avoid overhead in diff translations
+var (
+	maxTarget = big.NewFloat(0xFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFF)
+	minHash   = new(big.Float).Quo(new(big.Float).SetMantExp(big.NewFloat(1), 256), maxTarget)
```
really rolls right off the tongue

onemorebsmith commented on src/kaspastratum/stratum_server.go:
```diff
@@ -21,6 +22,9 @@ type BridgeConfig struct {
 	PrintStats      bool   `yaml:"print_stats"`
 	UseLogFile      bool   `yaml:"log_to_file"`
 	HealthCheckPort string `yaml:"health_check_port"`
+	BlockWaitTime   string `yaml:"block_wait_time"`
```
change this from a string to a time.Duration? It'll make life easier later on. You can parse it directly using flag.DurationVar

rdugan replied:
sure... was just following existing style, and am also new to go, so didn't dive in too deep.  Will take a look.

onemorebsmith commented on src/kaspastratum/stratum_test.go:
```diff
@@ -60,8 +60,12 @@ func TestHeaderSerialization(t *testing.T) {
 func TestPoolHzCalculation(t *testing.T) {
-	log.Println(shareValue)
-	log.Println(fixedDifficulty)
+	// TODO: figure out what we really want to test here.
```
Yeah this was a thing then I deleted it for whatever reason.

onemorebsmith reviewed (changes requested):
Thanks for doing this, just one comment around changing the block wait to a duration instead of a naked string everywhere

rdugan commented:
Ok, all three flags updated to appropriate types.  Sorry, should have been doing this from a branch - now i have a couple other minor updates included.  

 - Added total share diff counter to prometheus metrics, which is required for proper graphing of shares/hashrates w/o having to include mindiff in all queries (and will be critical if moving to vardiff)
 - Initialized all worker counters to 0, so rate functions can work properly w/ first data point
 - Added a couple dashboard json exports in the misc folder.  The first one should be a drop-in replacement for the existing dash, the second one ('_w_price') requires a new metric in prometheus to scrape KAS price, which i have not added

onemorebsmith commented:
@rdugan can you do a quick bullet-point writeup on everything you changed/added and I'll throw in in the release notes with a shoutout.

Also on how to use the new flags :)

Appreciate the work!

rdugan commented:
@onemorebsmith There's nothing really worth mentioning beyond the bullets in my original comment here.  Maybe just remove the diff instantiation comment, as it's unnecessary detail for the release notes?

Also, the config params are pretty extensively documented in the config.yaml file.
